src/lib/utils.py

```diff
@@ -108,13 +114,158 @@ def grouped_transform(
+    compat: bool = True,
+) -> DataFrame:
     return grouped_transform(
-        data, keys, lambda x: x.fillna(0).cumsum(), skip=skip, prefix=("total_", "new_")
+        data, keys, lambda x: x.fillna(0).cumsum(), skip=skip, prefix=prefix, compat=compat
+    )
+
+
+def stack_table(
+    data: DataFrame, index_columns: List[str], value_columns: List[str], stack_columns: List[str]
+) -> DataFrame:
+    """
+    Pivots a DataFrame's columns and aggregates the result as new columns with suffix. E.g.:
+
+    data:
+
+    idx piv val
+     0   A   1
+     0   B   2
+     1   A   3
+     1   B   4
+
+    stack_table(data, index_columns=[idx], value_columns=[val], stack_columns=[piv]):
+
+    idx val val_A val_B
+     0   3    1     2
+     1   7    3     4
+    """
+    output = data.drop(columns=stack_columns).groupby(index_columns).sum()
+
+    # Stash columns which are not part of the columns being indexed, aggregated or stacked
+    used_columns = index_columns + value_columns + stack_columns
+    stash_columns = [col for col in data.columns if col not in used_columns]
+    stash_output = data[stash_columns].copy()
+    data = data.drop(columns=stash_columns)
+
+    # Aggregate (stack) columns with respect to the value columns
+    for col_stack in stack_columns:
+        col_stack_values = data[col_stack].dropna().unique()
+        for col_variable in value_columns:
+            df = data[index_columns + [col_variable, col_stack]].copy()
+            df = df.pivot_table(
+                values=col_variable, index=index_columns, columns=[col_stack], aggfunc="sum"
+            )
+            column_mapping = {suffix: f"{col_variable}_{suffix}" for suffix in col_stack_values}
+            df = df.rename(columns=column_mapping)
+            transfer_columns = list(column_mapping.values())
+            output[transfer_columns] = df[transfer_columns]
+
+    # Restore the stashed columns, reset index and return
+    output[stash_columns] = stash_output
+    return output.reset_index()
+
+
+def age_group(age: int) -> str:
+    """
+    Categorical age group given a specific age, codified into a function to enforce consistency.
+    """
+    if age < 15:
+        return "children"
+    elif age < 65:
+        return "adult"
+    else:
+        return "elderly"
+
+
+def filter_index_columns(columns: List[str], index_schema: Dict[str, str]) -> List[str]:
+    """ Private function used to infer columns that this table should be indexed by """
+    index_columns = [col for col in columns if col in index_schema.keys()]
+    return index_columns + (["date"] if "date" in columns else [])
+
+
+def filter_output_columns(columns: List[str], output_schema: Dict[str, str]) -> List[str]:
+    """ Private function used to infer columns which are part of the output """
+    return [col for col in columns if col in output_schema.keys()]
+
+
+def infer_new_and_total(data: DataFrame, index_schema: Dict[str, str]) -> DataFrame:
```

Comment:
`grouped_diff` and `grouped_cumsum` are hot garbage I want to refactor ASAP, I just need to remove the calls from all parsing scripts (since these functions are no longer needed). I added tests for `infer_new_and_total` via ad9837bb0c2d5c2aa2aba3d0afb20934ffba66ad